### PR TITLE
Generalize data-parallel batching

### PR DESCRIPTION
**Pull request by rdraskicTT**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/23042
This PR continues the work of https://github.com/tenstorrent/tt-metal/pull/23129

### Problem description
vLLM can send varying batch sizes in the same run so we have to add support for this. Currently, it's assumed that `batch_size` is a multiple of `data_parallel` which is not always the case.

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/runs/15718294249)
- [x] [TG demo tests](https://github.com/tenstorrent/tt-metal/actions/runs/15718289965)
- [x] [T3K demo tests](https://github.com/tenstorrent/tt-metal/actions/runs/15685852329)
- [x] [vLLM nightly](https://github.com/tenstorrent/tt-metal/actions/runs/15718228990)
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: models/tt_transformers/tt/generator.py [hunk 1]
```diff
@@ -482,27 +483,26 @@ def prefill_forward(
...
-            cross_page_table = torch.chunk(cross_page_table, self.data_parallel, 0)
-
-        for group_user_id in range(batch_per_device):
-            for model_id in range(data_parallel):
-                user_id = group_user_id + model_id * batch_per_device
+            cross_page_table = torch.split(cross_page_table, batch_size_per_model)
 
+        user_id = 0
+        for model_id in range(self.data_parallel):
+            for group_user_id in range(batch_size_per_model[model_id]):
```
Comment (ipotkonjak-tt) [inline models/tt_transformers/tt/generator.py]: Will this increase prefill time? By swapping for loops you would now do prefill for all of the users on 1 model before moving to the second. This means that all of the devices from other DP groups are idle. The prefill is done for 1 user at a time so these cannot be overlapped.

If you restore the previous ordering of the loops where we iterate over models second we can do prefill in parallel on different devices. It is important not to mix up users from different DP groups in that case.

I'm curious about TTFT you are getting with the swapped loop order? Did you see a regression?

Comment (rdraskicTT) [reply models/tt_transformers/tt/generator.py]: That's a good catch, I wasn't aware that this could have an impact on performance. I'm checking TG demo tests and I don't see any significant difference in performance or prefill duration. Did you see this as an issue before or is this just theoretical expectation?

Comment (ipotkonjak-tt) [reply models/tt_transformers/tt/generator.py]: Just theoretical, I'd expect a difference in TTFT. But I guess if we just dispatch ops and then sync with tensor transfers at the end of all prefills we won't see that big of a diff.

**Review by ipotkonjak-tt — APPROVED** [review]

**Comment by rdraskicTT** [comment]
@skhorasganiTT Here I'm supporting different values for `self.model_args[i].max_batch_size`, but then the code might be unnecessarily complicated. Does this look ok to you or should we maybe assert somewhere that they are all equal and work with that assumption?

**Comment by skhorasganiTT** [comment]
> @skhorasganiTT Here I'm supporting different values for `self.model_args[i].max_batch_size`, but then the code might be unnecessarily complicated. Does this look ok to you or should we maybe assert somewhere that they are all equal and work with that assumption?

The max_batch_size per DP model is currently always the same, can assert and simplify a bit.

File: models/tt_transformers/tt/generator.py [hunk 2]
```diff
@@ -560,22 +564,23 @@ def decode_forward(
...
-        data_parallel = min(B, self.data_parallel)
-        batch_per_device = B // data_parallel
-        tokens = torch.chunk(tokens, self.data_parallel, 0)
-        start_pos = torch.chunk(start_pos, self.data_parallel, 0)
+        batch_size = tokens.shape[0]
+        batch_size_per_model = self._get_batch_size_per_model(batch_size)
+
+        tokens = torch.split(tokens, batch_size_per_model)
+        start_pos = torch.split(start_pos, batch_size_per_model)
+
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py]: I think for decode this might be overcomplicated/unnecessary since we can pad to the nearest DP and run the same batch on each submesh (also we currently pad to the global max batch in vLLM which is expected to be a multiple of DP). An assert for divisibility should probably be sufficient.

File: models/tt_transformers/tt/generator.py [hunk 3]
```diff
@@ -1165,6 +1170,15 @@ def _get_prefill_user_page_table(self, page_table, kv_cache, prefill_len):
...
         return page_table[:, :num_blocks]
 
+    def _get_batch_size_per_model(self, batch_size):
+        """
+        Returns a list of batch sizes for each model in the data parallel group.
+        """
+        max_batch_sizes = [0] + [self.model_args[i].max_batch_size for i in range(self.data_parallel)]
+        acc_batch_sizes = list(accumulate(max_batch_sizes))
+
+        return [max(0, min(batch_size - acc_batch_sizes[i], max_batch_sizes[i + 1])) for i in range(self.data_parallel)]
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py]: max_batch_size is the same for each DP model so this can be simplified a bit

File: models/tt_transformers/tt/generator.py (hunk 1 above)
Comment (skhorasganiTT) [reply models/tt_transformers/tt/generator.py]: +1, might be better to dispatch DP models in inner loop. And in that case, it would be better to modify `_get_batch_size_per_model` to allocate by looping over DPs one user at a time instead of fully allocate one DP then move to the next, though more complicated. Maybe it would just be simpler to pad to the nearest DP factor and discard/not read out padded users at the end.

File: models/tt_transformers/tt/generator.py (hunk 3 above)
Comment (rdraskicTT) [reply models/tt_transformers/tt/generator.py]: done

File: models/tt_transformers/tt/generator.py (hunk 1 above)
Comment (rdraskicTT) [reply models/tt_transformers/tt/generator.py]: Reverted back to this, can you please take another look and check if this is what you meant?

File: models/tt_transformers/tt/generator.py (hunk 2 above)
Comment (skhorasganiTT) [reply models/tt_transformers/tt/generator.py]: @rdraskicTT it looks like decode forward still has been modified, please check this comment

File: models/tt_transformers/tt/generator.py (hunk 1 above)
Comment (skhorasganiTT) [reply models/tt_transformers/tt/generator.py]: I see it's still using separate batch sizes per DP model, was it not preferred to pad the batch to the nearest DP factor and keep the old code?

Comment (rdraskicTT) [reply models/tt_transformers/tt/generator.py]: Sorry, I misunderstood what you meant the first time. I've reworked the code to pad to the nearest dp multiple.

File: models/tt_transformers/tt/generator.py [hunk 4]
```diff
@@ -57,32 +57,52 @@ def __init__(self, model, model_args, mesh_device, tokenizer=None, formatter=Non
...
 
         # Each model expected to run the same model, safe to use 1st vocab size
-        output_logits = torch.zeros(batch, 1, self.model_args[0].vocab_size)
-        prompt_lens = prompt_lens if prompt_lens is not None else torch.tensor([batch_seq_len] * batch)
-
-        data_parallel = min(batch, self.data_parallel)
-        batch_per_device = batch // data_parallel
+        output_logits = torch.zeros(padded_batch_size, 1, self.model_args[0].vocab_size)
+        prompt_lens = (
+            prompt_lens + [batch_seq_len] * (padded_batch_size - batch_size)
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py]: Can the padded prompts have length 1 instead of batch_seq_len (i.e. shortest possible instead of longest in batch)? Edit: may not need padded inputs, see comment about skipping prefill for padded users.

File: models/tt_transformers/tt/generator.py [hunk 5]
```diff
@@ -57,32 +57,52 @@ def __init__(self, model, model_args, mesh_device, tokenizer=None, formatter=Non
...
 
         if page_table is not None:
             assert isinstance(page_table, torch.Tensor), "page_table mush be torch.Tensor"
+            page_table = torch.cat(
+                [
+                    page_table,
+                    torch.zeros(padded_batch_size - batch_size, page_table.shape[1], dtype=torch.int32),
+                ],
+                dim=0,
+            )
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py]: Padding the page table with 0s may not be safe, since block 0 may get overwritten. I think -1 should be used instead, right @sraizada-tt? Edit: may not need padded inputs, see comment about skipping prefill for padded users.

File: models/tt_transformers/tt/generator.py [hunk 6]
```diff
@@ -482,33 +502,70 @@ def prefill_forward(
...
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py]: Same comment as above for page table padding. Edit: may not need padded inputs, see comment about skipping prefill for padded users.

File: models/tt_transformers/tt/generator.py [hunk 7]
```diff
@@ -57,32 +57,52 @@ def __init__(self, model, model_args, mesh_device, tokenizer=None, formatter=Non
...
+        if batch_size < padded_batch_size:
+            logger.info(f"Padding batch size from {batch_size} to {padded_batch_size}")
 
         # Each model expected to run the same model, safe to use 1st vocab size
-        output_logits = torch.zeros(batch, 1, self.model_args[0].vocab_size)
-        prompt_lens = prompt_lens if prompt_lens is not None else torch.tensor([batch_seq_len] * batch)
-
-        data_parallel = min(batch, self.data_parallel)
-        batch_per_device = batch // data_parallel
+        output_logits = torch.zeros(padded_batch_size, 1, self.model_args[0].vocab_size)
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py]: The `output_logits` should still have shape `batch_size` instead of `padded_batch_size` and we shouldn't return (or read/store) the outputs for the padded users. Also I realized, can we just skip calling `prefill_forward_single_user_text` for the padded users (and therefore no longer need the padded inputs)? Same comment for `prefill_forward`.

File: models/tt_transformers/tt/generator.py (hunk 4 above)
Comment (rdraskicTT) [reply models/tt_transformers/tt/generator.py]: na since we won't pad

File: models/tt_transformers/tt/generator.py (hunk 5 above)
Comment (rdraskicTT) [reply models/tt_transformers/tt/generator.py]: na since we won't pad

File: models/tt_transformers/tt/generator.py (hunk 6 above)
Comment (rdraskicTT) [reply models/tt_transformers/tt/generator.py]: na since we won't pad

File: models/tt_transformers/tt/generator.py [hunk 8]
```diff
@@ -1165,6 +1184,17 @@ def _get_prefill_user_page_table(self, page_table, kv_cache, prefill_len):
...
+    def _get_batch_size_per_model(self, batch_size):
+        """
+        Returns the maximum batch size per model and a list of batch sizes for each model.
+        """
+        max_batch_size_per_model = (batch_size + self.data_parallel - 1) // self.data_parallel
+
+        return max_batch_size_per_model, [
+            max(min(max_batch_size_per_model, batch_size - i * max_batch_size_per_model), 0)
+            for i in range(self.data_parallel)
+        ]
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py]: Could you please add a short description of the logic here for future readers?

**Review by skhorasganiTT — APPROVED** [review]
Thanks @rdraskicTT! Lgtm, but for sanity please also check that the existing DP t3k demo & tg demo CI tests still work.
